### DISCUSSION TITLE
자료구조 - 그래프 - 호현&일구&성은

### DISCUSSION THREAD

**Discussion by seongeun42** [Shortudy-10th/tech-for-developer#69] [질의 응답]
* 아래 2가지 질문에 답해주세요.
  * 그래프 IS A 트리. 참 거짓여부와 그 이유
  * 트리 IS A 그래프. 참 거짓여부와 그 이유
* 프로그램상에서 그래프를 구현한다면, 어떤 방법으로 구현할 수 있을까요?
* 대표적인 그래프 탐색법 두 개에 대해 간략하게 말해주세요.
* 완전 그래프와 연결 그래프의 차이점은 무엇일까요?
* 최소신장트리에 대해 설명해주세요.
* 다익스트라 알고리즘과 벨만포드 알고리즘에 대해 설명하고 두 알고리즘의 차이점을 설명해주세요
* 인접 행렬과 인접 리스트의 장단점은 무엇인가요?

**Comment by seongeun42** [comment]
### ❓️ 아래 2가지 질문에 답해주세요.
- 그래프 IS A 트리. 참 거짓 여부와 그 이유
- 트리 IS A 그래프. 참 거짓 여부와 그 이유

<br>

### 💬 답변
<blockquote>

<img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/d19f990b-0c4d-4353-8195-e9f7c2d48b3b" width="200">

<br>
<br>

- 그래프 IS A 트리. 참 거짓여부와 그 이유
    - **`False`**
    - 그래프란 여러 개의 노드와 노드를 연결하는 간선으로 구성된 자료 구조
    - 순환 그래프와 비순환 그래프 / 방향 그래프와 무방향 그래프로 나뉨
    - 노드 간의 경로가 여러 개 존재할 수 있음
    - 그래프가 트리보다 더 큰 범위이기 때문에 그래프가 트리라는 명제는 틀렸음
  <img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/bfde46b3-e872-42aa-b2a5-bf1b34d7619c" width="500">

<br> 

- 트리 IS A 그래프. 참 거짓여부와 그 이유
    - **`True`**
    - 트리는 그래프의 일종으로, 특정 조건을 만족하는 특수 그래프라고 보면 됨
    - 트리는 루트 노드가 있고, 노드 간의 경로가 하나만 존재하며, 부모, 자식 관계가 존재함
    - 사이클이 존재하지 않는 계층형 방향 그래프
    - 트리는 그래프의 일종이기 때문에 트리가 그래프라는 말도 맞음
    <img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/abeebbc2-950f-423d-aa67-8c6bc281c5ff" width="450">

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 프로그램 상에서 그래프를 구현한다면, 어떤 방법으로 구현할 수 있을까요?

<br>

### 💬 답변
<blockquote>

인접 행렬과 인접 리스트 두 가지 방법으로 구현할 수 있다.

- **인접 행렬** : 그래프를 2차원 배열을 사용해 구현하는 방법
- **인접 리스트** : 각 노드 별로 리스트를 만들어 연결되어 있는 노드를 해당 리스트에 저장하는 방법

<img src="https://github.com/Shortudy-10th/tech-for-developer/assets/64778589/548a78ac-aa44-4641-afa1-4d27999321f7" width="200">

```java
// 인접 행렬
int[][] arr = { { 0, 1, 1, 0, 0 },
		{ 1, 0, 0, 0, 0 },
		{ 1, 0, 0, 1, 0 },
		{ 0, 0, 1, 0, 1 },
		{ 0, 0, 0, 1, 0 } };

// 인접 리스트
int[][] list = { {},
		 { 2, 3 },
		 { 1 },
		 { 1, 4 },
		 { 3, 5 },
		 { 4 } };
```


</blockquote>

<br>

**Reply by seongeun42** [reply]
### ❓️ 인접 행렬과 인접 리스트의 장단점은 무엇인가요?

<br>

### 💬 답변
<blockquote>

* 인접 행렬
   * 장점
      * 2차원 배열에 모든 정점들의 간선 정보가 있기 때문에, 두 정점을 연결하는 간선을 조회할 때 $O(1)$ 시간복잡도로 가능
      * 정점(i)의 차수를 구할 때는 다음과 같이 인접행렬(M)의 i번째 행의 값을 모두 더하면 되므로 $O(n)$의 시간복잡도로 가능
      * 구현이 편함
   * 단점
      * 간선의 수와 무관하게 항상 $n^2$ 크기의 2차원 배열이 필요하므로 메모리 공간이 낭비
      * 그래프의 모든 간선의 수를 알아내려면 인접행렬 전체를 확인해야 하므로 $O(n^2)$의 시간이 소요
* 인접 리스트
   * 장점
      *  존재하는 간선만 관리하면 되므로 메모리 사용 측면에서 보다 효율적
      * 그래프의 모든 간선의 수를 알아내려면 각 정점의 헤더 노드부터 모든 인접리스트를 탐색해야 하므로 $O(n+e)$의 시간복잡도
   * 단점
      * 두 정점을 연결하는 간선을 조회하거나 정점의 차수를 알기 위해서는 정점의 인접 리스트를 탐색해야 하므로 정점의 차수만큼의 시간이 필요 -> $O(degree(v))$
      * 구현이 어려움

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 대표적인 그래프 탐색법 두 개에 대해 간략하게 말해주세요.

<br>

### 💬 답변
<blockquote>

* DFS와 BFS가 존재
* DFS   
![images_lucky-korma_post_30737a15-9adf-49a6-96a0-98c211cab1cc_R1280x0](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/5b0ded48-81c0-4e6e-a17a-7d813a14740b)

   * 특정 노드를 탐색한 후, 해당 노드의 이웃 노드를 다음으로 탐색
   * 이 과정을 해당 분기의 리프 노드가 나올때까지 반복
   * 리프 노드까지 탐색했으면, 이전 단계로 되돌아가 다른 이 노드에 대해 이를 반복
   * 시작 노드에서의 거리와 상관없이, 특정 분기에서 탐색할 수 있는 노드들을 먼저 방문
   * 스택 또는 재귀로 구현 가능
   * 이동 경로를 저장하는데 유리 & 탐색 대상 그래프가 엄청 클 경우 유리할 수 있음

* BFS
![images_lucky-korma_post_2112183b-bfcd-427e-8072-c9dc983180ba_R1280x0-2](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/46033113-037a-4503-87a4-efc15f3e1c32)

   * 특정 노드를 탐색한 후, 해당 노드의 이웃 노드들을 다음으로 탐색
   * 시작 노드에서 가까운 노드들을 먼저 방문하고, 거리가 먼 노드들을 나중에 방문
   * 큐를 사용해 구현 가능
   * 최단 거리를 찾는 데 유리

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 완전 그래프와 연결 그래프의 차이점은 무엇일까요?

<br>

### 💬 답변
<blockquote>

* 완전 그래프는 그래프 내에 있는 모든 정점간에 서로 간선이 존재하는 그래프
* 연결 그래프는 그래프 내에 있는 임의의 정점 간에 간선이 존재하는 그래프

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 최소신장트리에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

* 어떤 그래프에서의 MST는 다음 조건을 만족해야 함
   * 모든 정점이 MST에 포함돼야 한다
   * 사이클이 있으면 안 된다
   * 위의 조건을 만족하는 그래프들 중, 간선의 합이 최소인 그래프가 MST
* 한 그래프에서 MST는 여러 개가 나올 수 있다.
![다운로드](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/18c8ad9d-46d9-4157-ad05-18d28bfafb2a)


</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 그래프에서 MST를 어떻게 찾을 수 있을까요?

<br>

### 💬 답변
<blockquote>

크루스칼 알고리즘과 프림 알고리즘을 많이 사용한다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 크루스칼 알고리즘에 대해 설명해주세요.

<br>

### 💬 답변
<blockquote>

* 그리디 알고리즘의 일종
* 간선을 가중치를 기준으로 오름차순 정렬
* 사이클을 만들지 않는 선에서 그리디하게 간선을 선택
   * Union-Find 사용해 사이클 여부 판별 가능

* 예시
* 간선을 오름차순으로 정렬
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/5653cc9e-99a7-472b-a440-46908938602f)

* a-b 간선 선택
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/dbe856b2-d231-405e-a843-9495f7c03c72)

* a-d 간선 선택
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/dd02335f-9201-4d77-8fce-80d476614ec0)

* b-d를 선택 시 사이클이 생기므로 패스
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/890feb8d-a3e5-44e6-b0bc-80aa9fd1fbe4)

* b-c 간선 선택
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/37f2a79e-0ad9-4833-a6bc-64ea40243173)

* 간선의 개수가 노드의 개수 - 1개 이므로 종료.
   * Union-Find는 그래프 외적으로도 적용될 수 있는 알고리즘이라 자세히 옮기지 않음.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 프림 알고리즘에 대해 설명해주세요

<br>

### 💬 답변
<blockquote>

* 그리디 알고리즘의 일종
* 시작 노드를 MST 집합에 포함한 상태로 알고리즘 시작
* MST 집합에 속한 노드들과 가장 낮은 가중치로 연결된 노드를 선택.
   * 이때, 해당 노드가 이미 MST 집합에 포함돼 있다면 패스
   * 집합을 구성하는 자료구조는 딱히 말이 없는데, Set이나 배열로 잡는 듯
* MST집합의 크기가 노드의 개수와 같아진다면 종료

* 예시
* 초기상태 - A에서 시작
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/57723812-c9b9-4f3a-86e7-ad64ce11997e)

* A와 연결된 노드들 중 가장 가까운 노드는 C이므로 C를 집합에 추가
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/f0088913-dab3-4537-9d0d-b3d6db8ba9d1)

* A, C와 연결된 노드들 중 가장 가까운 노드는 B이므로 B를 추가
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/62d9c608-755b-4928-a27c-44de7b6f23e3)

* A, B, C와 연결된 노드들 중 가장 가까운 노드인 D 추가
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/62c936da-1346-4762-b45f-3e6fd2f776fd)

* A, B, C, D와 연결된 노드들 중 가장 가까운 노드인 E 추가
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/1addffb0-f96f-4ff2-a496-104ce6ce9017)

* A, B, C, D, E와 연결된 노드들 중 가장 가까운 노드인 F 추가
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/04fdc9a8-c01f-4d0a-81e2-f05414c5eaf9)

* MST 집합의 크기가 노드의 개수와 같으므로 종료

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 크루스칼 vs 프림

<br>

### 💬 답변
<blockquote>

* 프림은 정점 중심, 크루스칼은 간선 중심의 관점
* 프림은 시작점을 정하고, 가까운 정점들을 집합에 포함시켜 나가므로 그 과정에서 사이클이 발생하지 않음
   * 크루스칼은 간선을 추가하는 방식이므로 사이클 여부를 체크해줘야함
*  프림의 경우, 가장 가까운 거리의 정점을 찾는 부분에 어떤 자료구조를 쓰고, 어떻게 처리하느냐가 성능에 큰 영향을 줌
   * 크루스칼은 간선을 정렬하는 비용이 있음 - 보통 우선순위 큐를 쓴다
* 간선의 개수가 적으면 크루스칼이, 노드에 비해 간선이 엄청 많으면 프림이 유리할 것 같음

</blockquote>

<br>

**Comment by seongeun42** [comment]
### ❓️ 다익스트라 알고리즘과 벨만포드 알고리즘에 대해 설명하고 두 알고리즘의 차이점을 설명해주세요

<br>

### 💬 답변
<blockquote>

질문 흐름이 별로여서 꼬리질문에 제가 임의로 재구성하겠습니다.

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 다익스트라 알고리즘에 대해 설명해주세요

<br>

### 💬 답변
<blockquote>

* 한 노드에서 다른 모든 노드로의 거리를 구하는 알고리즘
* $V$를 노드의 개수, $E$는 한 노드의 주변 노드라고 할 때, 힙 트리를 사용하면 $O((V+E)logV)$의 시간복잡도를 가짐.
* 알고리즘 진행 과정
   1. 출발점으로부터의 최단거리를 저장할 배열을 만들고, 출발 노드를 0, 다른 노드들은 INF로 초기화
   2. 현재 노드를 나타내는 변수 A에 출발 노드의 번호를 저장
   3. A로부터 갈 수 있는 임의의 노드 B에 대해, **출발점에서 A까지의 거리 + A에서 B까지의 거리** 와 **출발점에서 B까지의 거리** 의 값을 비교
   4.  **출발점에서 A까지의 거리 + A에서 B까지의 거리**가 **출발점에서 B까지의 거리** 보다 작다면, 후자를 전자의 값으로 갱신
   5. A의 모든 이웃 노드에 대해 3~4의 과정을 반복
   6. A를 방문완료한 노드로 표시
   7. 미방문 노드들 중 출발점에서의 거리가 가장 짧은 노드를 찾아 다음 A로 설정
   8. 모든 노드를 방문하거나 더 이상 방문할 수 있는 노드가 없을 때 까지 위 과정을 반복
* 의사코드
```
function Dijkstra(Graph, Source):
 
     dist[source] ← 0                           // 소스와 소스 사이의 거리 초기화 --출발지와 출발지의 거리는 당연히 0--
      prev[source] ← undefined              // 출발지 이전의 최적 경로 추적은 없으므로 Undefined

     create vertex set Q                       //노드들의 집합 Q(방문되지 않은 노드들의 집합) 생성 시작

      for each vertex v in Graph:           // Graph안에 있는 모든 노드들의 초기화
          if v ≠ source:                          // V 노드가 출발지가 아닐 경우(출발지를 제외한 모든 노드!)
            dist[v] ← INFINITY             // 소스와 V노드 사이에 알려지지 않은 거리 --얼마나 먼지 모르니까-- = 무한값 ( 모든 노드들을 초기화하는 값)
            prev[v] ← UNDEFINED       // V노드의  최적경로 추적 초기화
        add v to Q                            //  Graph에 존재하고 방금 전 초기화된 V 노드를 Q(방문되지 않은 노드들의 집합)에 추가

//요약하자면, Graph에 존재하는 모든 노드들을 초기화한 뒤, Q에 추가함.
      
      while Q is not empty:                      // Q 집합이 공집합 아닐 경우, 루프 안으로!
          u ← vertex in Q with min dist[u]    // 첫번째 반복에서는 dist[source]=0이 선택됨. 즉, u=source에서 시작
          remove u from Q                         // U 노드를 방문한 것이므로 Q집합에서 제거
          
          for each neighbor v of u:           // U의 이웃노드들과의 거리 측정
              alt ← dist[u] + length(u, v)      // 출발지 노드 부터 계산된 U노드까지의 거리 + V에서 U의 이웃노드까지의 거리
                                                             //= source to U + V to U = source to U
             if alt < dist[v]:               // 방금 V노드까지 계산한 거리(alt)가 이전에 V노드까지 계산된 거리(dist[v])보다 빠른 또는 가까운 경우
                 dist[v] ← alt            //  V에 기록된 소스부터 V까지의 최단거리를 방금 V노드까지 계산한 거리로 바꿈
                  prev[v] ← u            // 제일 가까운 노드는 지금 방문하고 있는 노드(U)로 바꿈

     return dist[], prev[]       //계산된 거리값과 목적지를 리턴
```
* 진행과정 예시
   * S = 방문한 노드들의 집합
   * d[N] = A → N까지 계산된 최단 거리
   * Q = 방문하지 않은 노드들의 집합
   * 출발지는 A로 설정

* 초기 상태
   * 출발지는 0, 다른 노드들은 아직 확인이 안 됐기에 INF
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/8e50d00a-99f5-420b-9e89-a7e7a488ae66)

* A를 먼저 방문
   * A에서 이웃노드들과의 거리 계산 + 출발지와의 거리 갱신
   * 다음에 탐색할 노드는 거리가 가장 가까운 B
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/5a5ff7bc-6f5d-439a-ba73-7274e7e55d2b)

* B를 방문
   * 위 과정 반복
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/73ae1be0-e33a-4992-a0ae-0b397461509b)

* 일련의 과정에서 출발지와의 거리가 지속적으로 갱신됨
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/5cb559e2-396a-43d3-b67f-8e508547019b)

* 또 다른 예시
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/30398d46-4f3e-4f6e-977d-28e3f5977a0d)

* Q가 공집합이 되면 탐색 종료

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 다익스트라 알고리즘의 한계에 대해 설명해주세요

<br>

### 💬 답변
<blockquote>

* 그래프에 음수 사이클이 존재할 경우, 최단경로를 구할 수 없음
  * 음수 사이클이 존재하는 경우 해당 경로를 계산하는 과정에서 무한히 뺄 수 있게 되기 때문
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/5d3a0b54-1b48-4c9a-9e96-30097494bc56)

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 위 한계점을 극복할 수 있는 다른 알고리즘이 있을까요?
<br>

### 💬 답변
<blockquote>

* 음수 사이클이 존재하는 경우 **벨만포드 알고리즘**을 적용할 수 있음
* 알고리즘 진행 과정
   1. 시작 노드를 설정
   2. 시작 노드에서 각 다른 노드의 거리 값을 무한대로 설정하고 시작 노드를 0으로 설정
   3. 현재 노드의 모든 인접 노드를 탐색하며 기존에 저장된 인접 노드까지의 거리보다 현재 노드를 거치고 인접 노드에 도달하는 게 더 짧을 경우 값을 갱신
        - 시작 노드에서 특정 노드 사이의 모든 간선을 확인함
   4. 3을 모든 노드에 대해 수행
   5. 모든 노드에 3 - 4를 수행하고서 한 번 더 수행했을 때 또 거리가 갱신 된다면 음수 사이클이 존재함을 의미
       - 음수 사이클이 없다면 3 - 4를 수행한 후에는 최단 거리표에는 시작 노드에서 각 노드로 가는 최단 거리가 기록되어 있음. 음수 사이클이 존재한다면 한 번 더 도는 순간 최소값이 갱신되므로 존재한다는 뜻

* 예시
* 시작 노드 거리를 0으로, 나머지를 무한대로 초기화
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/22e772e4-0b15-4fb2-9ee2-212e79d2dc75)


* 인접 노드 값 확인 후 갱신
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/bbe91a1f-c96b-4054-b035-4229725e40c7)

* s의 인접노드인 a로 이동
   * 인접노드와 시작점의 거리 갱신

![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/790670c7-842e-4bad-9fdc-cff82ef0203c)

* a의 인접노드인 b로 이동 후 갱신 
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/624189f1-eb2b-423b-bab4-b42868d065e4)

* c 
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/18c36b1a-2dfd-42e1-9759-7229b852bff0)

* d
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/edf9ef24-9474-4f05-9412-bfa90118a072)

* e 
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/6fbbd1ef-98f9-4b14-9c9d-3852dd032f87)

* f
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/ecdd2f7d-ad47-4510-89df-82284b5e2ab1)

* g
![image](https://github.com/Shortudy-10th/tech-for-developer/assets/32262904/23b4e0d6-eb3d-45ce-8ccb-6efc6dfc84bf)

* 이 과정을 정점의 개수 - 1번 만큼 실행한다. 
   * 그 이후에도 거리의 최솟값들이 계속 갱신된다면, 음수 사이클이 있는 것
   * 예시의 경우 (c, d)와 (e, f)에 사이클이 존재하는데, 그 중 (e, f)는 음수값 절대값이 더 커 돌면 돌수록 거리가 감소하는 음수 사이클 -> 최단 경로를 구할 수 없는 경우임
* 음의 간선이 있어도 사용 가능한 알고리즘이지만, 기본적으로 모든 간선을 확인하므로 다익스트라보다 느림
  * 가중치에 음수값이 있는 경우, 그 중에서도 음수 사이클이 존재하는 상황에서만 사용해야 함

</blockquote>

<br>

**Reply by sikutisa** [reply]
### ❓️ 실제 현실 문제에서 다익스트라는 사용하기 적합한 알고리즘일까요?

<br>

### 💬 답변
<blockquote>

* 적합하지 않다.
   * 현실 공간에서의 장소나 거리 같은 것들을 모두 노드화시키기엔 그 수가 너무 많다.
   * 어떻게 한다고 해도, 현실에서는 최단 경로에 영향을 끼칠 수 있는 변수가 너무 많다.
*  순수 다익스트라를 적용하기 보단 인공지능을 접목하거나 A* 알고리즘 같은 변형 알고리즘을 사용하는 것이 더 용이
   * 위에 대한 자세한 내용은 주제 밖이므로 다루지 않음

</blockquote>

<br>
